### DISCUSSION TITLE
Should we allow registering duplicated route in Controller Registry?

### DISCUSSION THREAD

**Discussion by riflowth** [riflowth/springpress#14] [Q&A] [answered]
Because, if the controller has implemented too many routes, it may need to be separated classes for maintainability.

A real example from our tester repository
[MaintenaceLogController](https://github.com/CPE34-KMUTT/mod-tham-ngarn/blob/main/packages/backend/src/controllers/MaintenanceLogController.ts#L14)
[MaintenancePartController](https://github.com/CPE34-KMUTT/mod-tham-ngarn/blob/main/packages/backend/src/controllers/MaintenancePartController.ts#L14)

These controllers work on the main same thing, but their developers separate them into multiple classes for grouping a small work.
But doing this on [1.2.0](https://github.com/riflowth/springpress/releases/tag/1.2.0) will throw an error when registering those classes. [according to this implementation](https://github.com/riflowth/springpress/blob/1.2.0/lib/controllers/ControllerRegistry.js#L16-L18)

**Comment by XiaoXuxxxx** [comment, marked as answer]
>  if the controller has implemented too many routes, it may need to be separated classes for maintainability.

If that controller has too many routes. Then developer should create new controller, not the separate class with same controller

**Reply by riflowth** [reply]
So should we strict the developer by throwing an error when registering duplicated controllers in your opinion?
*according to @eltfshr  [answer](https://github.com/riflowth/springpress/discussions/14#discussioncomment-2691764)*

**Reply by XiaoXuxxxx** [reply]
I agree with @eltfshr, just show the warning instead of error

**Reply by riflowth** [reply]
Time to sum this up...

The developer `should create a new controller, not the separate class with the same controller mapping` as you said.

In one controller, too many routes may mean that the controller has too many purposes. so it should categorize into an individual controller that works on single responsibility instead.

According to @eltfshr answer

> I think it should be allowed but not encouraged. If able, you should avoid using duplicated routes. However, from the given example, maintainbility is better. 😏

I agree with @eltfshr, we should not encourage the developer to register duplicated routes to avoid code smells. 

Now the [ControllerRegistry#register](https://github.com/riflowth/springpress/blob/1.2.0/lib/controllers/ControllerRegistry.js#L16-L18) in 1.2.0 implementation, it throw an error when registering duplicated controller. So we will change the implementation to warn instead of throw an error. as @XiaoXuxxxx suggests.

*I will mark this as an answer.* 
Thanks to @XiaoXuxxxx and @eltfshr for participating in this discussion, your contribution will motivate the Springpress even further.

**Comment by eltfshr** [comment]
I think it should be allowed but not encouraged. If able, you should avoid using duplicated routes. However, from the given example, maintainbility is better. 😏

**Reply by riflowth** [reply]
Seem a lot of sense to not encouraged the developer to register duplicated controllers. But we will keep allowing developers to register duplicated controllers for some reasons that we have not yet faced.
*Your opinion will be taken as part of the Springpress development.* Thanks @eltfshr

**Comment by XiaoXuxxxx** [comment]
> A real example from our tester repository [MaintenaceLogController](https://github.com/CPE34-KMUTT/mod-tham-ngarn/blob/main/packages/backend/src/controllers/MaintenanceLogController.ts#L14) [MaintenancePartController](https://github.com/CPE34-KMUTT/mod-tham-ngarn/blob/main/packages/backend/src/controllers/MaintenancePartController.ts#L14)

These controllers don't work on the same thing. The reason that why these controller use the same controller is the limitation of accessible params from controller.

In this case MaintenanceLog's controller will be `/maintenance` and MaintenancePart's controller should be `/maintenance/:id/part` but the MaintenancePart's route will not able to access the `:id` from controller.

**Reply by XiaoXuxxxx** [reply]
I suggest that the route should be able to access the controller's params.
